### PR TITLE
Preparatory set of patches for `ALTER PROPERTY .. SET TYPE`

### PR DESCRIPTION
**Pull request by elprans**
See individual commits.

### CODE REVIEW

File: edb/schema/objects.py [hunk 1]
```diff
@@ -1460,24 +1460,27 @@ def init_delta_command(
...
         schema: s_schema.Schema,
+        context: sd.CommandContext,
         cmdtype: Type[sd.ObjectCommand_T],
         *,
         classname: Optional[sn.Name] = None,
         referrer: Optional[Object] = None,
         **kwargs: Any,
-    ) -> Tuple[sd.DeltaRoot, sd.ObjectCommand_T]:
-        root_cmd, parent_cmd = self.init_parent_delta_branch(
+    ) -> Tuple[sd.CommandGroup, sd.ObjectCommand_T, sd.ContextStack]:
```
Comment (msullivan) [inline edb/schema/objects.py:1480]: Might be worth adding a comment to init_delta_branch about this

File: edb/schema/delta.py [hunk 2]
```diff
@@ -2150,6 +2170,30 @@ def get_resolved_attribute_value(
 
         return value
 
+    def resolve_attribute_value(
```
Comment (msullivan) [inline edb/schema/delta.py:2173]: Can this share code with `get_resolved_attribute_value`?

File: edb/schema/delta.py [hunk 3]
```diff
@@ -3540,11 +3598,27 @@ def regular_cmd_from_ast(
...
+                and not astnode.value.elements
+            ):
                 # empty set
                 new_value = None
 
+            elif isinstance(astnode.value, qlast.TypeExpr):
+                if not isinstance(parent_op, QualifiedObjectCommand):
+                    raise AssertionError(
+                        'cannot determine module for derived compound type'
+                        ' parent operation is not a QualifiedObjectCommand'
```
Comment (msullivan) [inline edb/schema/delta.py:3612]: I'm not really sure how to parse this sentence; maybe some punctuation is intended after the first line?
(Not that it really matters, as an assertion)

File: edb/schema/inheriting.py [hunk 4]
```diff
@@ -170,13 +171,20 @@ def inherit_fields(
...
+                if isinstance(sav, sd.AlterObjectProperty):
+                    schema = self.scls.set_field_value(
+                        schema, field_name, result)
+                else:
+                    deferred_complex_ops.append(sav)
 
         schema = self._update_inherited_fields(
             schema, context, inherited_fields_update)
 
+        for op in deferred_complex_ops:
```
Comment (msullivan) [inline edb/schema/inheriting.py:185]: Why's this deferral needed? Add a comment?

File: edb/schema/pointers.py [hunk 5]
```diff
@@ -1477,16 +1462,28 @@ def _alter_begin(
...
+        context.current().enable_recursion = orig_rec
         scls = self.scls
 
+        vn = scls.get_verbosename(schema, with_parent=True)
+
         orig_target = scls.get_target(orig_schema)
         new_target = scls.get_target(schema)
 
+        if new_target is None:
+            # This will happen if `REST TYPE` was called
```
Comment (msullivan) [inline edb/schema/pointers.py:1477]: REST -> RESET

File: edb/schema/ordering.py [hunk 6]
```diff
@@ -482,6 +487,25 @@ def record_field_deps(
...
 
+    def write_dep_matrix(
+        dependent: str,
+        dependent_tags: Tuple[str, ...],
+        dependency: str,
+        dependency_tags: Tuple[str, ...],
+        *,
+        as_weak: bool = False,
+    ) -> None:
+        if as_weak:
```
Comment (msullivan) [inline edb/schema/ordering.py:498]: I'd push the `if as_week` into the loops?

(But: great helper function.)

File: edb/common/topological.py [hunk 7]
```diff
@@ -44,6 +44,7 @@ class DepGraphEntry(Generic[K, V, T]):
 
     item: V
     deps: MutableSet[K]
+    weak_deps: MutableSet[K]
```
Comment (msullivan) [inline edb/common/topological.py:47]: Could you add some comments on what the semantics of weak_deps are and how they are implemented?

**Review by msullivan — APPROVED** [review]
Looks good. Some nits and doc requests.

File: edb/schema/delta.py (hunk 3 above)
Comment (elprans) [reply edb/schema/delta.py:3612]: Yeah, I missed a colon there

File: edb/server/compiler/compiler.py [hunk 8]
```diff
@@ -874,9 +874,12 @@ def _compile_ql_migration(self, ctx: CompileContext, ql: qlast.Migration):
             mstate = mstate._replace(current_ddl=all_ddl)
             if debug.flags.delta_plan:
                 debug.header('Populate Migration DDL AST')
```
Comment (1st1) [inline edb/server/compiler/compiler.py:876]: This line is redundant

File: edb/schema/pointers.py [hunk 9]
```diff
@@ -482,13 +470,13 @@ def merge_targets(
...
                 vn = ptr.get_verbosename(schema)
+                t1_vn = t1.get_verbosename(schema)
                 t2_vn = t2.get_verbosename(schema)
                 raise errors.SchemaError(
                     f'cannot redefine {vnp} as {t2_vn}',
-                    details=(
-                        f'{vn} targets {t2_vn} that is not related '
-                        f'to a type found in this link in the parent type: '
-                        f'{t1.get_displayname(schema)!r}.'))
+                    details=f'{vn} is defined as {t1_vn} in a parent type, '
```
Comment (1st1) [inline edb/schema/pointers.py:477]: "in the parent type"? Also, I assume that `t1_vn` will include the parent type name, right?

**Review by 1st1 — APPROVED** [review]

File: edb/schema/pointers.py (hunk 9 above)
Comment (elprans) [reply edb/schema/pointers.py:477]: No it won't

Comment (elprans) [reply edb/schema/pointers.py:477]: And also, we are talking about _a_ parent type, because there may be several.

Comment (1st1) [reply edb/schema/pointers.py:477]: Can we include the actual name of the parent to the error message? Why can't we be specific here?

Comment (elprans) [reply edb/schema/pointers.py:477]: That would require some refactoring as we aren't passing the necessary context through the call stack. I can follow up with a cleanup later.

**Pull request merged**